### PR TITLE
fix(gemm): require per-batch MXFP8 BMM scales

### PR DESCRIPTION
**Pull request by wgu9**
## 📌 Description

This fixes incorrect CUTLASS `bmm_mxfp8` results on SM120 and SM121 when a batched matrix dimension is not aligned to 128.

The root cause is scale padding. CUTLASS reads each batch from its own 128-row scale tile, but the old code treated the full `B*M` or `B*N` range as one swizzled tensor. Later batches could then read scales from the wrong offsets.

This PR makes the CUTLASS path accept scales that are padded independently per batch. Callers can pass the scales as rank-preserving 3D tensors or as flattened 1D tensors when the layout is unambiguous. Ambiguous legacy 1D layouts are rejected because their shape is not enough to prove that the batch offsets are correct.

The SM120 binding now checks the same per-batch scale size that the kernel uses. The backend check also rejects unsupported `K` and `N` alignment before the FFI call.

The latest `main` already creates per-batch flattened scales in the trace template, so the earlier trace-specific changes are no longer needed in this diff.

## 🔍 Related Issues

Refs #3604

Related to #3457

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit`.
- [x] I have run the hooks on every changed file.

## 🧪 Tests

- [x] Tests cover non-aligned per-batch scales and legacy 1D rejection.
- [x] All pre-commit hooks pass after rebasing onto current `main`.

Before the rebase, I ran the targeted CUDA tests on an RTX 5090 with compute capability 12.0. The non-aligned cases, legacy-layout rejection, and aligned BMM checks all passed.

After the rebase, these checks pass locally.

```text
pre-commit run --files csrc/mxfp8_gemm_cutlass_sm120.cu flashinfer/gemm/gemm_base.py tests/gemm/test_bmm_mxfp8.py
python3 -m compileall -q flashinfer/gemm/gemm_base.py tests/gemm/test_bmm_mxfp8.py
git diff --check upstream/main...HEAD
```

This machine does not have PyTorch or CUDA, so the SM12x runtime tests need a CI rerun on an SM12x runner.

## Reviewer Notes

The scope is the SM120 and SM121 CUTLASS BMM path. The cuDNN path still requires flattened 1D scales.

### CODE REVIEW

**Comment by wgu9** [comment]
Addressed the bot review feedback in e918f702:

- Verified CodeRabbit's ambiguous 1D scale-length case (`b=2, m=100`) was a real correctness hole; `_prepare_bmm_mxfp8_cutlass_scale` now rejects ambiguous legacy 1D layouts before accepting by length, and the regression test covers both `m=17` and `m=100`.
- Verified Gemini's CUTLASS alignment point; `_cutlass_bmm_mxfp8_requirement` now filters invalid `K/N` alignment before dispatch.
- Applied Gemini's C++ simplification by using `b` directly for scale-size validation.
- Verified CodeRabbit's trace-template concern; `bmm_mxfp8_trace` now generates per-batch padded swizzled scales and has a metadata regression test.
- Applied the test skip helper nit for the new SM12x tests.

Validated locally on RTX 5090 / SM120:

```bash
python -m pytest tests/gemm/test_bmm_mxfp8.py::test_bmm_mxfp8_cutlass_rejects_combined_batch_scales tests/gemm/test_bmm_mxfp8.py::test_bmm_mxfp8_cutlass_non_aligned_m_per_batch_scales -q -s
python -m pytest tests/trace/test_fi_trace.py::test_bmm_mxfp8_trace_uses_per_batch_swizzled_scale_shapes -q
python -m pytest 'tests/gemm/test_bmm_mxfp8.py::test_bmm_mxfp8[False-cutlass-res_dtype0-True-input_dtype0-128-128-128-1]' 'tests/gemm/test_bmm_mxfp8.py::test_bmm_mxfp8[False-cutlass-res_dtype0-True-input_dtype0-128-128-128-16]' -q -s
python -m pre_commit run --all-files
```

All passed.

File: flashinfer/gemm/gemm_base.py [hunk 1]
```diff
@@ -8751,8 +8801,8 @@ def _cutlass_bmm_mxfp8_requirement(
     out: Optional[torch.Tensor] = None,
     backend: Literal["cudnn", "cutlass", "auto"] = "auto",
 ):
-    # SM120/121 CUTLASS MXFP8 only supports 1D swizzled scales.
-    if A_scale.ndim != 1 or B_scale.ndim != 1:
+    # SM120/121 CUTLASS MXFP8 supports flattened or rank-preserving swizzled scales.
+    if A_scale.ndim not in (1, 3) or B_scale.ndim not in (1, 3):
```
[the remark this answers is not in the document]
Comment (wgu9) [reply flashinfer/gemm/gemm_base.py:9146]: This guard is already present on the current branch: `_cutlass_bmm_mxfp8_requirement` now returns `False` when `A.shape[2] % 32 != 0` or `B.shape[2] % 32 != 0`, after the scale-rank check. That should keep unsupported K/N alignments from reaching the CUTLASS FFI path and allow backend fallback/selection instead.

**Comment by wgu9** [comment]
Ready for maintainer review from my side. I addressed the bot feedback in e918f702 and validated the ambiguous 1D scale-length case called out by CodeRabbit. Happy to run any additional RTX 5090-specific repro if useful.

**Comment by wgu9** [comment]
Rebased onto current `main` and resolved the `gemm_base.py` import conflict. The current branch includes the requested K/N alignment guard, ambiguous legacy-1D rejection, direct use of the existing batch variable in C++, SM12x test helper predicates, and a per-batch padded trace template plus trace regression coverage. Validation: Ruff check/format pass for all four touched Python files; `compileall` and `git diff --check upstream/main...HEAD` pass. CUDA runtime tests are left to CI on an SM12x-capable runner.

**Comment by wgu9** [comment]
Hi maintainers, I rebased this onto current `main` and resolved the conflicts.

While rebasing, I found one new interaction with the shared validation on `main`. It rejected rank-preserving 3D scales before the CUTLASS-specific flattening code could run. I updated the common check to allow 1D or 3D scales and kept cuDNN restricted to 1D. I also added CPU tests for that path and for both legacy 1D rejection cases.

The trace-specific part of the old diff is gone because current `main` already creates correctly padded per-batch flattened scales.

All pre-commit hooks pass on the three changed files. `compileall` and `git diff --check` pass too. This machine does not have PyTorch or CUDA, so the SM12x runtime tests still need CI.

Could someone with CI access please run the SM12x jobs when convenient? Thank you.

**Comment by wgu9** [comment]
Rebased onto current main and hardened the SM120 path after reproducing the original failure on an RTX 5090. At b=2, m=17, n=k=128, current main returned cosine 0.99929 for batch 0 and nonfinite output for batch 1. This branch keeps the scale tiles for each batch independent and rejects ambiguous combined-batch scales.

The audit also adds batch, dtype, device, output, and integer-range checks, raw binding validation and device guarding, plus a lock around workspace-size cache initialization.

Verification on RTX 5090:

- Explicit CUTLASS and auto matrix: 8 passed
- Boundary regression tests: 13 passed
- Raw binding and two-thread stream stress: passed
- pre-commit: passed
- git diff --check: passed

**Comment by wgu9** [comment]
The trace input path already follows the new scale contract on this rebased branch. `_bmm_mxfp8_init` quantizes each batch matrix independently, then concatenates the padded 128x4 scale buffers. `test_bmm_mxfp8_trace_init_quantizes_each_batch` checks the per-batch calls and physical scale lengths.

Rechecked the trace coverage: 3 passed.

**Comment by wgu9** [comment]
Correction to my earlier note: independent quantization followed by flattening still fails for b=2, M=100, N=K=160. At that shape, the per-batch buffer and the legacy combined-batch buffer have the same length, so CUTLASS rejects the flattened input.

The trace path now keeps the layouts separate. cuDNN uses combined-batch 1D scales. CUTLASS uses rank-preserving per-batch scales. The initializer selects the layout for the current CUDA architecture.

Validated on RTX 5090:

- Trace and production cases: 15 passed
- MXFP8 boundary cases: 13 passed
- Trace contract tests: 35 passed
- Template tests: 6 passed
- Raw binding and two-thread, two-stream stress: passed
- pre-commit and git diff check: passed
